File: Daily Coding Problem/Barclays/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        
        
        stack  =[]
        
        for i in s:
            if len(stack)==0:
                stack.append(i)
                
            elif i==')' and stack[-1]=='(':
                stack.pop()
            elif i==']' and stack[-1]=='[':
                stack.pop()
            elif i=='}' and stack[-1]=='{':
                stack.pop()
            else:
                stack.append(i)
        
        if len(stack)==0:
            return True
            
        return False

File: Data Structure/4.Stack/Valid Parentheses.java
// https://leetcode.com/problems/valid-parentheses/
import java.util.*;

class Solution {
    public boolean isValid(String s) {
    
        Stack<Character> stack = new Stack<>();
        
        char first = s.charAt(0);
        if (s.length()<2) return false;
        
        else if ( first == '}' || first == ']' || first == ')') return false;
        
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);
            if (ch == '(' || ch=='[' || ch=='{' ){
                stack.push(ch);
            }
            else if ( ch== '}' || ch==']' || ch==')') {
                char peek = '0';
                if (!stack.empty()) peek = stack.peek();
                if (ch== '}' && peek == '{' ){
                    stack.pop();
                }
                else if (ch== ']' && peek == '['){
                    stack.pop();
                }
                else if (ch == ')' && peek== '('){
                    stack.pop();
                }
                else{
                    return false;
                }
                
                
                
            }
        }
        
        
        if (!stack.empty()) return false;
        return true;
    }
}